Refactor activity

Dimi20cen opened the pull request:
Aim is to improve readability mainly by
- Use more descriptive naming for variables and functions
- Breaking up large functions
- Adding global variables

Other changes
- Incorporate score_path inside scorescreen
- Fill space that was unused by pygame with black color

@chimosky @sourabhaa, I `git cherry-pick ca712d3`, is that close to what you suggested me to do?

Dimi20cen committed:
Refactor activity

Aim is to improve readability mainly by
- Use more descriptive naming for variables and functions
- Breaking up large functions
- Adding global variables

Other changes
- Incorporate score_path inside scorescreen
- Fill space that was unused by pygame with black color

chimosky commented on elements.py:
```diff
@@ -1,84 +1,58 @@
-                self.shape = 1
-            else:
-                self.shape = 2
+        if (self.x_to_draw in (OBJECTS_X_POSITIONS[0],
```
The outer parentheses isn't needed.

Dimi20cen replied:
Yep, fixed

chimosky commented on elements.py:
```diff
@@ -1,84 +1,58 @@
+            game.objectlist.remove(self)
+
+    def objects_interactions(self, game):
+        if (self.object_rect.colliderect(self.car1_rect) or
+                self.object_rect.colliderect(self.car2_rect)):
```
The outer parentheses is redundant.

Dimi20cen replied:
Yep, fixed

chimosky commented on elements.py:
```diff
@@ -1,84 +1,58 @@
+                game.score += 1
+                game.objectlist.remove(self)
+
+        # Missed a coin(circle)
+        if (self.object in (game.red_circle, game.blue_circle) and
+                self.y_to_draw >= 640):
+            game.miss.play(0)
+            game.coin_miss = True
```
The comment here is redundant and the outer parentheses too.

Dimi20cen replied:
Yep, fixed

chimosky commented on main.py:
```diff
@@ -20,235 +20,278 @@
+from welcomescreen import Welcomescreen
+from scorescreen import Scorescreen
+
+# Enable use of activity with `python3 game.py`
+os.environ['SUGAR_BUNDLE_ID'] = 'org.sugarlabs.2cars'
```
Avoid setting activity bundle_id here, running the game with `python3 game.py` works as expected.

Dimi20cen replied:
Without this line, when in scorescreen and try to restart the game I get 
```
 Traceback (most recent call last):
  File "/home/dim/Documents/GitHub/2-cars-activity/main.py", line 289, in <module>
    game.run()
  File "/home/dim/Documents/GitHub/2-cars-activity/main.py", line 279, in run
    self.update_game_state()
  File "/home/dim/Documents/GitHub/2-cars-activity/main.py", line 259, in update_game_state
    self.handle_collision()
  File "/home/dim/Documents/GitHub/2-cars-activity/main.py", line 185, in handle_collision
    screen = Scorescreen(self)
  File "/home/dim/Documents/GitHub/2-cars-activity/scorescreen.py", line 56, in __init__
    get_activity_root(), 'data', 'score.pkl')
  File "/usr/local/lib/python3.10/dist-packages/sugar3/activity/activity.py", line 1565, in get_activity_root
    activity_root = env.get_profile_path(os.environ['SUGAR_BUNDLE_ID'])
  File "/usr/lib/python3.10/os.py", line 680, in __getitem__
    raise KeyError(key) from None
KeyError: 'SUGAR_BUNDLE_ID'
```

chimosky replied:
It works perfectly on my end, @jriyyya can you test?

chimosky commented on main.py:
```diff
@@ -20,235 +20,278 @@
+        pygame.mixer.init()
+        self.load_assets()
+
+    def run(self):
+        self.initialize()
+        clock = pygame.time.Clock()
+        self.running = True
+        self.music.play(-1)
+        self.screen = pygame.display.get_surface()
+
+        while self.running:
+            self.render()
+            if self.state == WELCOME:
+                screen = Welcomescreen(self)
+                screen.run(self)
+                self.state = PLAY
+
+            self.handle_events()
+            self.update_game_state()
+
+            pygame.display.update()
+            clock.tick(FPS)
+        pygame.quit()
+        sys.exit()
```
If all the functions used here are defined later in the code then it makes sense to have those definitions come before this.

Dimi20cen replied:
Yep, fixed

chimosky commented on main.py:
```diff
@@ -20,235 +20,278 @@
+
+            pygame.display.update()
+            clock.tick(FPS)
+        pygame.quit()
+        sys.exit()
```
A return instead of `pygame.quit` and `sys.exit` should be fine.

Dimi20cen replied:
Fixed

chimosky commented on main.py:
```diff
@@ -20,235 +20,278 @@
+        sys.exit()
+
+    def load_assets(self):
+        # Load Images
```
Redundant comment.

Dimi20cen replied:
Fixed

chimosky commented on main.py:
```diff
@@ -20,235 +20,278 @@
+        self.red_square_image = pygame.image.load(RED_SQUARE_PATH)
+        self.blue_square_image = pygame.image.load(BLUE_SQUARE_PATH)
+
+        # Load Music
```
Redundant comment.

Dimi20cen replied:
Fixed

chimosky commented on main.py:
```diff
@@ -20,235 +20,278 @@
+        self.music = pygame.mixer.Sound(SOUND_MUSIC_PATH)
+        self.scoresound = pygame.mixer.Sound(SOUND_SCORE_PATH)
+
+    # Scale images according to screen size, only happens once
```
Redundant comment, although the comment says it's according to screen size but the scaling is littered with hard coded values.

Dimi20cen replied:
Fixed

chimosky commented on main.py:
```diff
@@ -20,235 +20,278 @@
+        score_object = self.font3.render(_(str(self.score)), 1, (WHITE))
+        self.screen.blit(score_object, (780, 20))
+
+    def handle_events(self):
+        while Gtk.events_pending():
+            Gtk.main_iteration()
+        for event in pygame.event.get():
+            self.handle_input(event)
+
+    def handle_input(self, event):
+        if event.type == pygame.QUIT:
+            self.running = False
+        if event.type == pygame.KEYDOWN:
+            if event.key == pygame.K_LEFT and not self.is_left_pressed:
+                self.left_moved = True
+                self.is_left_pressed = True
+            elif event.key == pygame.K_RIGHT and not self.is_right_pressed:
+                self.right_moved = True
+                self.is_right_pressed = True
+        elif event.type == pygame.KEYUP:
+            if event.key == pygame.K_LEFT:
+                self.is_left_pressed = False
+            elif event.key == pygame.K_RIGHT:
+                self.is_right_pressed = False
```
This should be one function, not two separate functions.

Dimi20cen replied:
Agree, fixed

chimosky commented on main.py:
```diff
@@ -20,235 +20,278 @@
+            else:
+                while True:
+                    soon_to_draw = Element(self)
+                    if (soon_to_draw.x_to_draw < 530 and
+                       soon_to_draw.object != self.lastleft.object):
```
Redundant outer parentheses.

Dimi20cen replied:
Fixed

chimosky commented on main.py:
```diff
@@ -20,235 +20,278 @@
+            if self.last.x_to_draw < 530:
+                while True:
+                    soon_to_draw = Element(self)
+                    if (soon_to_draw.x_to_draw > 530 and
+                       soon_to_draw.object != self.lastright.object):
```
Redundant outer parentheses.

Dimi20cen replied:
fixed

chimosky commented on main.py:
```diff
@@ -20,235 +20,278 @@
+                self.right_angle += ANGLE_SPEED
+
+    def update_car_positions(self):
+        # Left car
```
Redundant comment.

Dimi20cen replied:
Fixed

chimosky commented on main.py:
```diff
@@ -20,235 +20,278 @@
+                if self.leftcar_x <= 390:
+                    self.left_moved = False
+                    self.left = not self.left
+        # Right car
```
Redundant comment.

Dimi20cen replied:
Fixed

chimosky commented on scorescreen.py:
```diff
@@ -21,98 +21,112 @@
+        return scaled_image
+
+    def read_max_score(self):
+        if not os.path.exists(self.score_path):
+            open(self.score_path, 'w+')
```
Why do this when self.score_path isn't assigned to a variable here but just opened and won't be closed as it doesn't have a reference, plus you use a context manager below to access the file and that handles creation if the file doesn't exist.

Dimi20cen replied:
Fixed

chimosky commented on scorescreen.py:
```diff
@@ -21,98 +21,112 @@
-            pygame.quit()
-            sys.exit()
+            clock.tick(FPS)
+        pygame.quit()
+        sys.exit()
```
Please get rid of these too and use a return.

Dimi20cen replied:
Fixed

chimosky commented on welcomescreen.py:
```diff
@@ -20,92 +20,92 @@
 
-class welcomescreen:
+WELCOMESCREEN_PATH = os.path.join("assets", "welcomescreen.png")
+PLAY_BUTTON_PATH = os.path.join("assets/play.png")
+RULESCREEN_PATH = os.path.join("assets/rules.png")
```
Make these consistent with the other uses.

Dimi20cen replied:
Fixed

chimosky commented on welcomescreen.py:
```diff
@@ -20,92 +20,92 @@
-                if event.type == pygame.KEYDOWN and (event.key == 276 or event.key == 275):
-                    return
+                    self.running = False
+                if (event.type == pygame.KEYDOWN and
+                   (event.key == pygame.K_LEFT or
+                       event.key == pygame.K_RIGHT)):
+                    if ruleflag == 0:
+                        ruleflag = 1
```
This isn't readable, you can check for a `KEYDOWN` event first then check if it's the right or left key that was pressed separately.

Dimi20cen replied:
Fixed

chimosky commented on welcomescreen.py:
```diff
@@ -20,92 +20,92 @@
-    g = welcomescreen()
-    g.run(gameDisplay)
+            clock.tick(FPS)
+        pygame.quit()
+        sys.exit()
```
Please remove these, the activity itself handles exit events so no need to call these within the activity.

Dimi20cen replied:
Fixed

Dimi20cen committed:
Improve consistency and remove redundancy

- Change order of functions in main.py to improve readability
- Remove redundant comments and outer parentheses
- Remove redundant use of pygame.quit() and sys.exit()

Dimi20cen committed:
Import main instead of game

Left out from when main.py was named game.py

Dimi20cen committed:
Refactor high score read/write methods

- Remove unnecessary file opening in `read_max_score`
- Utilize context managers for safer file handling
- Simplify conditional checks for file existence and size

Dimi20cen commented:
@chimosky Fixed reviews

chimosky commented:
Reviewed e06169d, not tested.

For future commit messages, leave a line before writing your commit message body as that helps for better formatting.

Dimi20cen commented:
Yep, sorry about the no new line, I use GitHub Desktop and sometimes forgot to add it, thanks for the review

chimosky commented:
Could you rebase master on your branch?

Dimi20cen commented:
Sure, no problem, done.

chimosky commented on activity.py:
```diff
@@ -81,6 +81,3 @@ def sound_control(self, button):
             button.set_tooltip(_('Sound'))
-
-    def get_preview(self):
-        return self._pygamecanvas.get_preview()
```
Why'd you remove the `get_preview`?

Dimi20cen replied:
Sorry about that, fixed

Dimi20cen committed:
Add get_preview in activity.py

Method get_preview was removed by commit, 589d91f0bec5755c9331bcef9d056e4dcf13324e, for no apparent reason, so it's added back to the activity.py

chimosky commented on game.py:
```diff
@@ -20,239 +20,276 @@
+from scorescreen import Scorescreen
+
+# Enable use of activity with `python3 game.py`
+os.environ['SUGAR_BUNDLE_ID'] = 'org.sugarlabs.2cars'
```
Please remove this.

Dimi20cen replied:
Hm, I guess that's an issue on my end then. Fixed

chimosky replied:
Bundle ID isn't supposed to be set within an activity, the code for getting [bundle id](https://github.com/sugarlabs/sugar-toolkit-gtk3/blob/master/src/sugar3/activity/activity.py#L674-L679) tells you where it's supposed to be set.

Dimi20cen committed:
Remove unnecessary environment var setting

chimosky commented on game.py:
```diff
@@ -20,239 +20,273 @@
 
 
 class Game:
+<<<<<<< HEAD:game.py
 
+=======
+>>>>>>> 34c7954 (Refactor activity):main.py
```
Forget to remove these?

Dimi20cen replied:
Leftover from the rebase, I thought I couldn't continue rebase untill all of those were resolve, aparently i was wrong. Anyways fixed

chimosky commented:
In 589d91f you switched to main but the file main doesn't exist and neither does it on master.

What's the reason behind the switch?

Dimi20cen committed:
Remove lines from rebase

Dimi20cen committed:
Fix main file import in activity.py

Dimi20cen commented:
Sorry about that, I originally changed the name of the game.py file to main.py, trying to follow convention. But in this PR I didn't change the naming. Just fixed the wrong import in the activity.py.

chimosky commented:
> Sorry about that, I originally changed the name of the game.py file to main.py, trying to follow convention. But in this PR I didn't change the naming. Just fixed the wrong import in the activity.py.

Some activities have game.py and some have main.py, nothing wrong with the name of the file.

Conventions matter more with code style than names of files.
